server/region_handler.go

```diff
@@ -0,0 +1,486 @@
+			RegionEpoch: meta.RegionEpoch,
+		}
+
+	}
+	return regions, nil
+}
+
+// ServeHTTP handles request of list a table's regions.
+func (rh TableRegionsHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	// parse params
+	params := mux.Vars(req)
+	dbName := params[pDBName]
+	tableName := params[pTableName]
+	// check and prepare tools
+	tool, err := rh.prepare()
+	if err != nil {
+		rh.writeError(w, err)
+		return
+	}
+	// get table's schema.
+	table, err := tool.infoSchema.TableByName(model.NewCIStr(dbName), model.NewCIStr(tableName))
+	if err != nil {
+		rh.writeError(w, err)
+		return
+	}
+	tableID := table.Meta().ID
+
+	// for record
+	startKey, endKey := tablecodec.GetTableHandleKeyRange(tableID)
+	recordRegionIDs, err := tool.regionCache.ListRegionIDsInKeyRange(tool.bo, startKey, endKey)
+	if err != nil {
+		rh.writeError(w, err)
+		return
+	}
+	recordRegions, err := rh.getRegionsMetaWithRegionIds(recordRegionIDs)
+	if err != nil {
+		rh.writeError(w, err)
+		return
+	}
+
+	// for indices
+	indices := make([]IndexRegions, len(table.Indices()))
+	for i, index := range table.Indices() {
+		indexID := index.Meta().ID
+		indices[i].Name = index.Meta().Name.String()
+		indices[i].ID = indexID
+		startKey, endKey := tablecodec.GetTableIndexKeyRange(tableID, indexID)
+		rIDs, err := tool.regionCache.ListRegionIDsInKeyRange(tool.bo, startKey, endKey)
+		if err != nil {
+			rh.writeError(w, err)
+			return
+		}
+		indices[i].Regions, err = rh.getRegionsMetaWithRegionIds(rIDs)
+		if err != nil {
+			rh.writeError(w, err)
+			return
+		}
+	}
+
+	tableRegions := &TableRegions{
+		TableName:     tableName,
+		TableID:       tableID,
+		Indices:       indices,
+		RecordRegions: recordRegions,
+	}
+
+	rh.writeData(w, tableRegions)
+}
+
+// ServeHTTP handles request of get region by ID.
+func (rh RegionHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	// parse and check params
+	params := mux.Vars(req)
+	regionIDInt, err := strconv.ParseInt(params[pRegionID], 0, 64)
+	if err != nil {
+		rh.writeError(w, err)
+		return
+	}
+	regionID := uint64(regionIDInt)
+
+	// check and prepare tools
+	tool, err := rh.prepare()
+	if err != nil {
+		rh.writeError(w, err)
+		return
+	}
+
+	// locate region
+	region, err := tool.regionCache.LocateRegionByID(tool.bo, regionID)
+	if err != nil {
+		rh.writeError(w, err)
+		return
+	}
+
+	frameRange, err := NewRegionFrameRange(region)
+	if err != nil {
+		rh.writeError(w, err)
+		return
+	}
+
+	// create RegionDetail from RegionFrameRange
+	regionDetail := &RegionDetail{
+		RegionID: regionID,
+		StartKey: region.StartKey,
+		EndKey:   region.EndKey,
+	}
+	// Since we need a database's name for each frame, and a table's database name can not
+	// get from table's ID directly. Above all, here do dot process like
+	// 		`for id in [frameRange.firstTableID,frameRange.endTableID]`
+	// on [frameRange.firstTableID,frameRange.endTableID] is small enough.
+	for _, db := range tool.infoSchema.AllSchemas() {
```

Comment:
Since we need a database's name for each frame, and a table's database name can not get from table's ID directly. So here do dot process like
 ```
for id in [frameRange.firstTableID,frameRange.endTableID]
       ...
```
 on [frameRange.firstTableID,frameRange.endTableID] is small enough.

Above all, it seems iterate all databases and tables cannot be avoided. With `infoSchema` already have all databases with tables in memory, If we iterate databases by name, and iterate tables in each databases separately in the following way, to reduce a little little memory?

```
	for _, db := range tool.infoSchema.AllSchemaNames() {
		for _, table := range tool.infoSchema.SchemaTables(model.NewCIStr(db)) {
			start, end := frameRange.getIndexRangeForTable(table.Meta().ID)
			regionDetail.addTableInRange(db, table.Meta(), start, end)
		}
	}
```